README.md
rvw
===

Converts data sets in Vowpal Wabbit's sparse format to sparse lists and dense data frames in R. Useful in conjunction with downsampling big data in Vowpal Wabbit format for prototyping models in R and conducting exploratory data anlaysis.

### Demo
[http://alexkchavez.shinyapps.io/rvw-demo/](http://alexkchavez.shinyapps.io/rvw-demo/)

### Getting Started
From R, `install.packages("devtools")` and then install the version `>= 1.9.3` of `data.table` using the instructions below (copied from [https://github.com/Rdatatable/data.table](https://github.com/Rdatatable/data.table)):

    # check and update to latest version on CRAN
    update.packages()
    
    # try latest development version from GitHub
    # ( Windows users should first install Rtools
    #   http://cran.r-project.org/bin/windows/Rtools/ )
    require(devtools)
    install_github("data.table", "Rdatatable")
    
    # if you get pdflatex or texi2dvi errors during installation, want a 
    # quick way out and don't mind skipping building vignettes:
    install_github("data.table", "Rdatatable", build_vignettes=FALSE)
    
Finally, install rvw: `install_github("rvw", "alexkchavez")`.

### Usage

    > require(rvw)
    > vw <- c(
      "1 1.0 |MetricFeatures:3.0 height:1.5 length:2.0 |Has stripe |Other Legs:4.0 IsAnimal",
      "1 1.0 zebra|MetricFeatures:3.0 height:1.5 length:2.0 |Has stripe |Other Legs:4.0 IsAnimal",
      "1 1.0 zebra|MetricFeatures:1.5 length:2.0 |Has a white stripe |Other Legs:4.0 IsAlive",
      "1 2 'tag|a:2 b:3",
      "0 |f:.23 sqft:.25 age:.05 2006"
    )

    > fromVw(vw, dense = T)
      label importance   tag MetricFeatures_height MetricFeatures_length Has_stripe Other_Legs Other_IsAnimal Has_a Has_white
    1     1          1  <NA>                   4.5                     6          1          4              1    NA        NA
    2     1          1 zebra                   4.5                     6          1          4              1    NA        NA
    3     1          1 zebra                    NA                     3          1          4             NA     1         1
    4     1          2  'tag                    NA                    NA         NA         NA             NA    NA        NA
    5     0          1  <NA>                    NA                    NA         NA         NA             NA    NA        NA
      Other_IsAlive a_b f_sqft  f_age f_2006
    1            NA  NA     NA     NA     NA
    2            NA  NA     NA     NA     NA
    3             1  NA     NA     NA     NA
    4            NA   6     NA     NA     NA
    5            NA  NA 0.0575 0.0115   0.23

    > fromVw(vw, dense = F)
    [[1]]
      label importance  tag MetricFeatures_height MetricFeatures_length Has_stripe Other_Legs Other_IsAnimal
    1     1          1 <NA>                   4.5                     6          1          4              1
    
    [[2]]
      label importance   tag MetricFeatures_height MetricFeatures_length Has_stripe Other_Legs Other_IsAnimal
    1     1          1 zebra                   4.5                     6          1          4              1
    
    [[3]]
      label importance   tag MetricFeatures_length Has_a Has_white Has_stripe Other_Legs Other_IsAlive
    1     1          1 zebra                     3     1         1          1          4             1
    
    [[4]]
      label importance  tag a_b
    1     1          2 'tag   6
    
    [[5]]
      label importance  tag f_sqft  f_age f_2006
    1     0          1 <NA> 0.0575 0.0115   0.23


NEWS
rvw 0.1
=======
  
* Initial release. `fromVw` functions converts a vector of Vowpal Wabbit strings to a list of sparse R data frames or to a dense data frame.

* Added tests for most functions.

* A TODO is to add conversion from R data frames to Vowpal Wabbit.

tests/test-all.R
library(testthat)
test_check("rvw")


tests/testthat/test_from_vw.R
context("Convert VW format to data frame")

# Good formats
vw <- c(
  "1 1.0 |MetricFeatures:3.0 height:1.5 length:2.0 |Has stripe |Other Legs:4.0 IsAnimal",
  "1 1.0 zebra|MetricFeatures:3.0 height:1.5 length:2.0 |Has stripe |Other Legs:4.0 IsAnimal",
  "1 1.0 zebra|MetricFeatures:1.5 length:2.0 |Has a white stripe |Other Legs:4.0 IsAlive",
  "1 2 'tag|a:2 b:3",
  "0 |f:.23 sqft:.25 age:.05 2006"
)

# List of expected R objects
expected <- {
  
  df1 <- data.frame(
    label = "1",
    importance = 1.0,
    tag = as.character(NA),
    MetricFeatures_height = 3.0 * 1.5, 
    MetricFeatures_length = 3.0 * 2.0,
    Has_stripe = 1.0,
    Other_Legs = 4.0,
    Other_IsAnimal = 1.0,
    stringsAsFactors = F
  )
  
  df2 <- {
    df2 <- df1
    df2$tag = "zebra"
    df2
  }
  
  df3 <- data.frame(
    label = "1",
    importance = 1.0,
    tag = "zebra",
    MetricFeatures_length = 1.5 * 2.0,
    Has_a = 1.0,
    Has_white = 1.0,
    Has_stripe = 1.0,
    Other_Legs = 4.0,
    Other_IsAlive = 1.0,
    stringsAsFactors = F
  )
  
  df4 <- data.frame(
    label = "1",
    importance = 2.0,
    tag = "'tag",
    a_b = 2.0 * 3.0,
    stringsAsFactors = F
  )
  
  df5 <- data.frame(
    label = "0",
    importance = 1.0,
    tag = as.character(NA),
    f_sqft = 0.23 * 0.25,
    f_age = 0.23 * 0.05,
    f_2006 = 0.23,
    stringsAsFactors = F
  )
  
  list(df1, df2, df3, df4, df5)
}

# Bad formats
vw.bad <- c(
  "1 2 'tag |a:2 b:3",
  "1 2 tag |a:2 b:3",
  "1 2 'second_house |f:.18 sqft:.15 age:.35 1976",
  "0 1 0.5 'third_house |f:.53 sqft:.32 age:.87 1924"
)

# Tests

test_that("Parses vowpal wabbit string to sparse R format correctly", {
  for (i in length(expected)) expect_equal(fromVw(vw[i], dense = F), expected[i])
})

test_that("Parses vowpal wabbit string to dense R format correctly", {
  for (i in length(expected)) expect_equal(fromVw(vw[i], dense = T), expected[[i]])
})

test_that("Parses vowpal wabbit vector to sparse R format correctly",
  expect_equal(fromVw(vw, dense = F), expected)
)

test_that("Parses vowpal wabbit data to dense R format correctly",
  expect_equal(fromVw(vw, dense = T), as.data.frame(data.table::rbindlist(expected, fill = T)))
)

test_that("Cannot parse bad data", {
  for (x in vw.bad) expect_error(fromVw(x), "Could not parse")
})


tests/testthat/test_from_vw_helpers.R
context("Convert VW format to R list")

test_that("vwNamespaceToList cannot parse bad data", {
  expect_error(vwNamespaceToList('Has'), "Could not parse")
  expect_error(vwNamespaceToList(' Has Stripe'), "Could not parse")
})

test_that("vwNamespaceToList parses good data", {
  expect_equal(vwNamespaceToList("Has stripe "), list(Has_stripe = 1.0))
  expect_equal(vwNamespaceToList("Other Legs:4.0 IsAnimal"), list(Other_Legs = 4.0, Other_IsAnimal = 1.0))
  expect_equal(
    vwNamespaceToList("MetricFeatures:3.0 height:1.5 length:2.0 "),
    list(MetricFeatures_height = 3.0 * 1.5, MetricFeatures_length = 3.0 * 2.0)
  )
})

test_that("vwToList cannot parse bad data", {
  expect_error(vwNamespacesToList('a b:2.0 |Has '), "Could not parse")
  expect_error(vwNamespacesToList('a b:2.0 | Has Stripe'), "Could not parse")
})

test_that("vwToList parses good data", {
  expect_equal(vwNamespacesToList("Has stripe"), list(list(Has_stripe = 1.0)))
  expect_equal(
    vwNamespacesToList("Has stripe |Other Legs:4.0 IsAnimal"),
    list(list(Has_stripe = 1.0), list(Other_Legs = 4.0, Other_IsAnimal = 1.0))
  )
})


R/from_vw.R
#' Convert Vowpwal Wabbit Dataset to List or Data Frame
#' 
#' Convert a vector of strings in Vowpal Wabbit format to a list of sparse data frames or
#' to a dense data frame.
#' 
#' Vowpal Wabbit strings are in the format:
#' 
#' [Label] [Importance [Tag]]|Namespace Features |Namespace Features ... |Namespace Features
#' 
#' where 
#' 
#' \itemize{
#'   \item Label is an real number representing the dependent variable. Because examples can be unlabeled, Label is optional.
#'   \item Importance is the non-negative importance weight, and defaults to 1.0 if unspecified.
#'   \item Tag is an optional tag for the example, and defaults to NA if unspecified.
#'   \item Namespace=String[:Value] is a namespace string optionally followed by a float representing the global weight for features in the namespace. If the weight is omitted, 1.0 is substituted.
#'   \item Features=(String[:Value] )* is a sequence of feature names optionally followed by a float representing the feature weight. If the weight is omitted, 1.0 is substituted.
#' }
#' 
#' Conversion to R objects occurs by concatenating the namespace string with each feature
#' name in that namespace, and multiplying the namespace global weight with the feature
#' weight.
#' 
#' Additional details on the Vowpal Wabbit format can be found here:
#' \url{https://github.com/JohnLangford/vowpal_wabbit/wiki/Input-format}.
#'
#' @param x vector of Vowpal Wabbit strings.
#' @param dense whether to return the data in dense format as a data frame, or in a sparse
#'   format as a list. Defaults to false.
#' @param keepvw whether to include the original Vowpal Wabbit string in the output. 
#'   Defaults to false.
#' @return Data frame of examples including columns label, importance, tag, feature 
#'   values, and optionally the raw input string.
#' @examples
#' vw <- c(
#'   "1 2 'tag|a:2 b:3",
#'   "0 |f:.23 sqft:.25 age:.05 2006"
#' )
#'   
#' fromVw(x = vw, dense = FALSE)
#' # [[1]]
#' #   label importance  tag a_b
#' # 1     1          2 'tag   6
#' #
#' # [[2]]
#' #   label importance  tag f_sqft  f_age f_2006
#' # 1     0          1 <NA> 0.0575 0.0115   0.23
#'
#' fromVw(x = vw, dense = TRUE)
#' #   label importance  tag a_b f_sqft  f_age f_2006
#' # 1     1          2 'tag   6     NA     NA     NA
#' # 2     0          1 <NA>  NA 0.0575 0.0115   0.23
#'
#' @export
fromVw <- function(x, dense = FALSE, keepvw = FALSE) {
  regex <- '^(?<label>[^ ]+) (?:(?<importance>[^ ]+) (?:(?<tag>[^ ]+))?)?\\|(?<features>[^ ].*)'

  sparse.features <- lapply(x, function(vw) {
    regex.result <- regexpr(regex, vw, perl = TRUE)
    parsed <- as.list(parseRegexResults(vw, regex.result))
    l <- {
      l <- if (keepvw) parsed else parsed[c("label", "importance", "tag")]
      l$importance <- if (l$importance == "") 1.0 else as.numeric(l$importance)
      l$tag <- if (l$tag == "") NA_character_ else l$tag
      c(l, vwNamespacesToList(parsed$features))
    }
    data.frame(l, stringsAsFactors = FALSE)
  })
  
  if (!dense) {
    sparse.features
  } else {
    as.data.frame(data.table::rbindlist(sparse.features, fill = TRUE))
  }
}

#' Convert the Vowpal Wabbit 'Namespace Features |Namespace Features ... |Namespace
#' Features' to a list
#' 
#' @param x Namespace Features |Namespace Features ... |Namespace Features
#' @return list of lists with names equal to the namespace prepended to feature names and 
#'   values equal to the product of the namespace weight and feature weights.
#' @examples
#'   # Returns list(namespace1_feature1 = 4.0, namespace1_feature2 = 6.0,
#'   # namespace2_feature1 = 1.0, namespace2_feature3 = 1.0)
#'   vwNamespacesToList('namespace1:2.0 feature1:2.0 feature2:3.0 |namespace2 feature1 feature3')
#' @keywords internal
#' @export
vwNamespacesToList <- function(x) {
  feature.sections <- unlist(strsplit(x, split = ' |', fixed = TRUE))
  lapply(feature.sections, vwNamespaceToList)
}

#' Convert Vowpal Wabbit 'Namespace Features ' section to a list
#' 
#' Converts Vowpal Wabbit 'Namespace Features ' section by concatenating the namespace
#' string with each feature name and multiplying the namespace global weight with the
#' feature weight. Namespace=String[:Value] is a namespace string optionally followed by a
#' float representing the global weight for features in the namespace, and 
#' Features=(String[:Value] )* is a sequence of feature names optionally followed by a 
#' float representing the feature weight. If a weight is omitted, 1.0 is substituted.
#' 
#' @param x Vowpal Wabbit 'Namespace Features ' section where Namespace=String[:Value] and
#'   Features=(String[:Value] )*. If a weight is omitted, 1.0 is substituted.
#' @return list with names equal to the namespace prepended to feature names and values 
#'   equal to the product of the namespace weight and feature weights.
#' @examples
#'   # Returns list(namespace1_feature1 = 4.0, namespace1_feature2 = 6.0)
#'   vwNamespaceToList('namespace1:2.0 feature1:2.0 feature2:3.0 ')
#'   # Returns list(namespace2_feature1 = 1.0, namespace2_feature3 = 1.0)
#'   vwNamespaceToList('namespace2 feature1 feature3 ')
#' @keywords internal
#' @export
vwNamespaceToList <- function(x) {
  
  regex <- '(?<feature>[^:]+)(?::(?<weight>.*))?'
  
  .normalize <- function(fw) {
    if (length(fw) <= 1) stop(paste0("Could not parse: ", fw))
    regex.result <- regexpr(regex, fw, perl = TRUE)
    m <- {
      m <- do.call(rbind, parseRegexResults(fw, regex.result))
      m[m[, "weight"] == "", "weight"] <- "1.0"
      m
    }
    namespace.feature <- paste0(m[1, "feature"], "_", m[-1, "feature"])
    l <- {
      adjusted.weight <- as.list(as.numeric(m[1, "weight"]) * as.numeric(m[-1, "weight"]))
      names(adjusted.weight) <- namespace.feature
      adjusted.weight
    }
    l
  }
  
  .normalize(unlist(strsplit(x, split = " ", fixed = TRUE)))
}
